### PR TITLE
Use nullptr (C++11) instead of zero (0) as the null pointer constant

### PR DESCRIPTION
**Pull request by practicalswift**
Use `nullptr` instead of zero (0) as the null pointer constant.

The road towards `nullptr` (C++11) is split into two PRs:
* `NULL` → `nullptr` is handled in PR #10483 (scripted)
* `0` → `nullptr` is handled in PR #10645 (manual, this PR)

By using the C++11 keyword `nullptr` we are guaranteed a prvalue of type `std::nullptr_t`.

For a more thorough discussion, see "A name for the null pointer: nullptr" (Sutter &
Stroustrup), http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2431.pdf

### CODE REVIEW

**Comment by jtimon** [comment]
utACK a0fbff2c8c0621d2ef4759078b3c690f1c33748b

**Review by gmaxwell — COMMENTED** [review]
Concept ACK

**Comment by practicalswift** [comment]
Rebased!

**Comment by practicalswift** [comment]
@laanwj Thanks for merging the `NULL → nullptr` PR (#10483). This PR handles the case `0 → nullptr` and has now been rebased on top of `master`.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/10645/commits/f71710dad454f5afe431479d0aea578b36d68206

**Review by dcousens — APPROVED** [review]

File: src/tinyformat.h [hunk 1]
```diff
@@ -898,7 +898,7 @@ class FormatListN : public FormatList
 // Special 0-arg version - MSVC says zero-sized C array in struct is nonstandard
 template<> class FormatListN<0> : public FormatList
 {
-    public: FormatListN() : FormatList(0, 0) {}
+    public: FormatListN() : FormatList(nullptr, 0) {}
```
Comment (jnewbery) [inline src/tinyformat.h]: There may be an argument to not making style changes to `tinyformat.h`, which occasionally pulls in changes from upstream. Don't know what @laanwj's view is on this?

**Review by jnewbery — COMMENTED** [review]
Tested ACK f71710dad454f5afe431479d0aea578b36d68206. Looks good @practicalswift . Thanks for doing this!

One open question for @laanwj

File: src/tinyformat.h (hunk 1 above)
Comment (laanwj) [inline src/tinyformat.h]: Agree, we should generally minimize changes to tinyformat.h, and if it is just style it just adds burden on us with no clear gain. 

If you get upstream to accept this it's okay with me. But I think they want to keep compatiblity with older C++.

**Comment by practicalswift** [comment]
@jnewbery @laanwj Good point! The change to `src/tinyformat.h` is no longer included in this PR.

**Pull request merged**